### PR TITLE
[CARBONATA-3605] Remove global dictionary feature

### PR DESCRIPTION
**Pull request by jackylk**
Since CarbonData supports local dictionary, it is a more scalable feature than global dictionary. This PR remove the writing of the dictionary and modify all testcases

 - [ ] Any interfaces changed?
 
 - [ ] Any backward compatibility impacted?
 
 - [ ] Document update required?

 - [ ] Testing done
        Please provide details on 
        - Whether new unit test cases have been added or why no new tests are required?
        - How it is tested? Please attach test report.
        - Is it a performance related change? Please attach the performance test report.
        - Any additional information to help reviewers in testing this change.
       
 - [ ] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA.

### CODE REVIEW

**Comment by CarbonDataQA1** [comment]
Build Failed with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/1124/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.2, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1134/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/1115/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/1118/

**Comment by CarbonDataQA1** [comment]
Build Failed with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/1127/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.2, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1137/

**Comment by MakKind** [comment]
retest this please

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1153/

**Comment by CarbonDataQA1** [comment]
Build Failed with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/1143/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/1134/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1195/

**Comment by CarbonDataQA1** [comment]
Build Failed with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/1185/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/1176/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/1177/

**Comment by CarbonDataQA1** [comment]
Build Failed with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/1186/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1196/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/1179/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1198/

**Comment by CarbonDataQA1** [comment]
Build Failed with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/1188/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/1203/

**Comment by CarbonDataQA1** [comment]
Build Failed with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/1212/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1222/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1227/

**Comment by CarbonDataQA1** [comment]
Build Failed with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/1217/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/1208/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/1214/

**Comment by CarbonDataQA1** [comment]
Build Failed with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/1224/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1233/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/1223/

**Comment by CarbonDataQA1** [comment]
Build Failed with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/1233/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1242/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/1267/

**Comment by CarbonDataQA1** [comment]
Build Failed with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/1277/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1288/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/1271/

**Comment by CarbonDataQA1** [comment]
Build Failed with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/1281/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1292/

**Comment by ajantha-bhat** [comment]
I suggest we can keep 2 PR, one for load flow and one for query flow removal as changes will be huge.

**In load flow we can consider removing dictionary server (one pass)** also as it works with global  dictionary

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/1293/

**Comment by CarbonDataQA1** [comment]
Build Failed with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/1302/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1313/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/1297/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1317/

**Comment by CarbonDataQA1** [comment]
Build Failed with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/1306/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/1300/

**Comment by CarbonDataQA1** [comment]
Build Failed with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/1309/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1322/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/1304/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1327/

**Comment by CarbonDataQA1** [comment]
Build Failed with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/1314/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/1317/

File: docs/configuration-parameters.md [hunk 1]
```diff
@@ -90,7 +90,6 @@ This section provides the details of all the configurations required for the Car
...
 | carbon.dictionary.chunk.size | 10000 | CarbonData generates dictionary keys and writes them to separate dictionary file during data loading. To optimize the IO, this configuration determines the number of dictionary keys to be persisted to dictionary file at a time. **NOTE:** Writing to file also serves as a commit point to the dictionary generated. Increasing more values in memory causes more data loss during system or application failure. It is advised to alter this configuration judiciously. |
 | dictionary.worker.threads | 1 | CarbonData supports Optimized data loading by relying on a dictionary server. Dictionary server helps to maintain dictionary values independent of the data loading and there by avoids reading the same input data multiples times. This configuration determines the number of concurrent dictionary generation or request that needs to be served by the dictionary server. **NOTE:** This configuration takes effect when ***carbon.options.single.pass*** is configured as true. Please refer to *carbon.options.single.pass*to understand how dictionary server optimizes data loading. |
```
Comment (QiangCai) [inline docs/configuration-parameters.md]: remove this

File: docs/ddl-of-carbondata.md [hunk 2]
```diff
@@ -83,7 +83,6 @@ CarbonData DDL statements are documented here,which includes:
 
 | Property                                                     | Description                                                  |
 | ------------------------------------------------------------ | ------------------------------------------------------------ |
-| [DICTIONARY_INCLUDE](#dictionary-encoding-configuration)     | Columns for which dictionary needs to be generated           |
 | [NO_INVERTED_INDEX](#inverted-index-configuration)           | Columns to exclude from inverted index generation            |
 | [INVERTED_INDEX](#inverted-index-configuration)              | Columns to include for inverted index generation             |
```
Comment (QiangCai) [inline docs/ddl-of-carbondata.md]: no need

File: hadoop/src/main/java/org/apache/carbondata/hadoop/readsupport/impl/DictionaryDecodeReadSupport.java [hunk 3]
```diff
@@ -60,11 +59,9 @@ public void initialize(CarbonColumn[] carbonColumns,
         Cache<DictionaryColumnUniqueIdentifier, Dictionary> forwardDictionaryCache = cacheProvider
             .createCache(CacheType.FORWARD_DICTIONARY);
         dataTypes[i] = carbonColumns[i].getDataType();
-        String dictionaryPath = carbonTable.getTableInfo().getFactTable().getTableProperties()
-            .get(CarbonCommonConstants.DICTIONARY_PATH);
         dictionaries[i] = forwardDictionaryCache.get(new DictionaryColumnUniqueIdentifier(
             carbonTable.getAbsoluteTableIdentifier(),
-            carbonColumns[i].getColumnIdentifier(), dataTypes[i], dictionaryPath));
+            carbonColumns[i].getColumnIdentifier(), dataTypes[i]));
```
Comment (QiangCai) [inline hadoop/src/main/java/org/apache/carbondata/hadoop/readsupport/impl/DictionaryDecodeReadSupport.java]: use CarbonRowReadSupport instead of this class

File: integration/spark-common/src/main/scala/org/apache/spark/sql/catalyst/CarbonDDLSqlParser.scala [hunk 4]
```diff
@@ -767,78 +766,19 @@ abstract class CarbonDDLSqlParser extends AbstractCarbonSparkSQLParser {
...
 
     // by default consider all String cols as dims and if any dictionary include isn't present then
     // add it to noDictionaryDims list. consider all dictionary excludes/include cols as dims
     fields.foreach { field =>
-      if (dictExcludeCols.exists(x => x.equalsIgnoreCase(field.column))) {
-        noDictionaryDims :+= field.column
-        dimFields += field
-      } else if (dictIncludeCols.exists(x => x.equalsIgnoreCase(field.column))) {
-        dimFields += field
-      } else if (field.dataType.get.toUpperCase.equals("TIMESTAMP") &&
```
Comment (QiangCai) [inline integration/spark-common/src/main/scala/org/apache/spark/sql/catalyst/CarbonDDLSqlParser.scala]: need check this change, maybe exclude timestamp column, it will be non-dict column

File: processing/src/main/java/org/apache/carbondata/processing/loading/converter/impl/RowConverterImpl.java [hunk 5]
```diff
@@ -98,70 +82,21 @@ public void initialize() throws IOException {
...
-              configuration.getParentTablePath(), isConvertToBinary,
+          .createFieldEncoder(fields[i], i, nullFormat, isEmptyBadRecord, isConvertToBinary,
               (String) configuration.getDataLoadProperty(
                   CarbonLoadOptionConstants.CARBON_OPTIONS_BINARY_DECODER));
       fieldConverterList.add(fieldConverter);
     }
     CarbonTimeStatisticsFactory.getLoadStatisticsInstance()
         .recordLruCacheLoadTime((System.currentTimeMillis() - lruCacheStartTime) / 1000.0);
-    fieldConverters = fieldConverterList.toArray(new FieldConverter[fieldConverterList.size()]);
+    fieldConverters = fieldConverterList.toArray(new FieldConverter[0]);
```
Comment (QiangCai) [inline processing/src/main/java/org/apache/carbondata/processing/loading/converter/impl/RowConverterImpl.java]: why size is zero?

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1341/

**Comment by CarbonDataQA1** [comment]
Build Failed with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/1330/

File: conf/carbon.properties.template [hunk 6]
```diff
@@ -85,7 +85,6 @@ carbon.major.compaction.size=1024
 ##Min max feature is added to enhance query performance. To disable this feature, make it false.
 #carbon.enableMinMax=true
 
-######## Global Dictionary Configurations ########
```
[the remark this answers is not in the document]
Comment (jackylk) [reply conf/carbon.properties.template]: they are still required

File: core/src/main/java/org/apache/carbondata/core/metadata/schema/table/CarbonTable.java [hunk 7]
```diff
@@ -704,8 +704,6 @@ public boolean isHivePartitionTable() {
   public AbsoluteTableIdentifier getAbsoluteTableIdentifier() {
     AbsoluteTableIdentifier absoluteTableIdentifier =
         tableInfo.getOrCreateAbsoluteTableIdentifier();
-    absoluteTableIdentifier.setDictionaryPath(
-        tableInfo.getFactTable().getTableProperties().get(CarbonCommonConstants.DICTIONARY_PATH));
     return absoluteTableIdentifier;
```
[the remark this answers is not in the document]
Comment (jackylk) [reply core/src/main/java/org/apache/carbondata/core/metadata/schema/table/CarbonTable.java]: fixed

File: docs/configuration-parameters.md (hunk 1 above)
Comment (jackylk) [reply docs/configuration-parameters.md]: fixed

File: docs/ddl-of-carbondata.md (hunk 2 above)
Comment (jackylk) [reply docs/ddl-of-carbondata.md]: inverted index is not removed, it is required

File: processing/src/main/java/org/apache/carbondata/processing/loading/converter/impl/RowConverterImpl.java (hunk 5 above)
Comment (jackylk) [reply processing/src/main/java/org/apache/carbondata/processing/loading/converter/impl/RowConverterImpl.java]: It is recommended to give 0 size array to `.toArray` func in java

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/1323/

**Comment by CarbonDataQA1** [comment]
Build Failed with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/1333/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1346/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/1324/

**Comment by CarbonDataQA1** [comment]
Build Failed with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/1334/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1347/

File: core/src/main/java/org/apache/carbondata/core/datastore/page/encoding/DefaultEncodingFactory.java [hunk 8]
```diff
@@ -77,7 +76,6 @@ public ColumnPageEncoder createEncoder(TableSpec.ColumnSpec columnSpec, ColumnPa
   private ColumnPageEncoder createEncoderForDimension(TableSpec.DimensionSpec columnSpec,
       ColumnPage inputPage) {
     switch (columnSpec.getColumnType()) {
-      case GLOBAL_DICTIONARY:
       case DIRECT_DICTIONARY:
```
[the remark this answers is not in the document]
Comment (jackylk) [reply core/src/main/java/org/apache/carbondata/core/datastore/page/encoding/DefaultEncodingFactory.java]: Yes, if we can think of a good name, we can consider it in another PR

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/1325/

**Comment by CarbonDataQA1** [comment]
Build Failed with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/1335/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1348/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/1326/

**Comment by CarbonDataQA1** [comment]
Build Failed with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/1336/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1349/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/1328/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/1338/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1351/

File: conf/carbon.properties.template (hunk 6 above)
Comment (ajantha-bhat) [inline conf/carbon.properties.template]: **please search "gobaldict" and "global dict", some more places need to handle I guess**

File: core/src/main/java/org/apache/carbondata/core/cache/CacheProvider.java [hunk 9]
```diff
@@ -135,11 +131,11 @@ public static CacheProvider getInstance() {
   private void createDictionaryCacheForGivenType(CacheType cacheType) {
```
Comment (ajantha-bhat) [inline core/src/main/java/org/apache/carbondata/core/cache/CacheProvider.java]: now it is not dictionaryCache, we can rename it.

Comment (ajantha-bhat) [reply core/src/main/java/org/apache/carbondata/core/cache/CacheProvider.java]: also can rename **dictionaryCacheAlreadyExists** API

File: conf/carbon.properties.template (hunk 6 above)
Comment (jackylk) [reply conf/carbon.properties.template]: The dictionary read path is not removed, I will do it in another PR

File: core/src/main/java/org/apache/carbondata/core/cache/dictionary/AbstractColumnDictionaryInfo.java [hunk 10]
```diff
@@ -17,7 +17,6 @@
 
 package org.apache.carbondata.core.cache.dictionary;
 
-import java.nio.charset.Charset;
```
Comment (ajantha-bhat) [inline core/src/main/java/org/apache/carbondata/core/cache/dictionary/AbstractColumnDictionaryInfo.java]: why not remove this whole class ?

File: core/src/main/java/org/apache/carbondata/core/localdictionary/dictionaryholder/DictionaryByteArrayWrapper.java [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+
+import org.apache.carbondata.core.util.ByteUtil;
+
+import net.jpountz.xxhash.XXHash32;
+
+/**
+ * class that holds the byte array and overrides equals and hashcode method which
+ * will be useful for object comparison
+ */
+public class DictionaryByteArrayWrapper {
```
Comment (ajantha-bhat) [inline core/src/main/java/org/apache/carbondata/core/localdictionary/dictionaryholder/DictionaryByteArrayWrapper.java]: same class present in **package org.apache.carbondata.core.cache.dictionary**, need to remove duplicate code from there ?

Comment (ajantha-bhat) [reply core/src/main/java/org/apache/carbondata/core/localdictionary/dictionaryholder/DictionaryByteArrayWrapper.java]: better move it, instead of add and delete

File: core/src/main/java/org/apache/carbondata/core/scan/filter/resolver/resolverinfo/visitor/DictionaryColumnVisitor.java [hunk 12]
```diff
@@ -55,26 +54,22 @@ public void populateFilterResolvedInfo(ColumnResolvedFilterInfo visitableObj,
...
+        // first check of surrogate key for null value is already added then
+        // no need to add again otherwise result will be wrong in case of exclude filter
+        // this is because two times it will flip the same bit
+        if (!resolvedFilterObject.getExcludeFilterList()
+            .contains(CarbonCommonConstants.MEMBER_DEFAULT_VAL_SURROGATE_KEY)) {
+          resolvedFilterObject.getExcludeFilterList()
+              .add(CarbonCommonConstants.MEMBER_DEFAULT_VAL_SURROGATE_KEY);
         }
-      } catch (QueryExecutionException e) {
-        throw new FilterUnsupportedException(e);
```
Comment (ajantha-bhat) [inline core/src/main/java/org/apache/carbondata/core/scan/filter/resolver/resolverinfo/visitor/DictionaryColumnVisitor.java]: why this change ? Want to throw QueryExecutionException exception itself, but this method many place throws  FilterUnsupportedException itself

File: core/src/main/java/org/apache/carbondata/core/util/CarbonLoadStatisticsImpl.java [hunk 13]
```diff
@@ -314,9 +299,7 @@ private void printDicGenStatisticsInfo() {
     double loadCsvfilesToDfTime = getLoadCsvfilesToDfTime();
     LOGGER.info("STAGE 1 ->Load csv to DataFrame and generate" +
             " block distinct values: " + loadCsvfilesToDfTime + "(s)");
-    double dicShuffleAndWriteFileTotalTime = getDicShuffleAndWriteFileTotalTime();
-    LOGGER.info("STAGE 2 ->Global dict shuffle and write dict file: " +
-            dicShuffleAndWriteFileTotalTime + "(s)");
+    LOGGER.info("STAGE 2 ->Global dict shuffle and write dict file (deprecated): 0(s)");
```
Comment (ajantha-bhat) [inline core/src/main/java/org/apache/carbondata/core/util/CarbonLoadStatisticsImpl.java]: better to remove this log itself

File: hadoop/src/main/java/org/apache/carbondata/hadoop/readsupport/impl/DirectDictionaryReadSupport.java [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+import org.apache.carbondata.core.metadata.datatype.DataTypes;
+import org.apache.carbondata.core.metadata.schema.table.CarbonTable;
+import org.apache.carbondata.core.metadata.schema.table.column.CarbonColumn;
+import org.apache.carbondata.hadoop.readsupport.CarbonReadSupport;
+
+/**
+ * A read support implementation to return Object array
+ * Only does direct dictionary (date/timestamp) conversion
+ */
+public class DirectDictionaryReadSupport<T> implements CarbonReadSupport<T> {
```
Comment (ajantha-bhat) [inline hadoop/src/main/java/org/apache/carbondata/hadoop/readsupport/impl/DirectDictionaryReadSupport.java]: This is not called anywhere now ? 
also when handle read flow, can remove **CarbonDictionaryDecodeReadSupport**

File: integration/hive/src/main/java/org/apache/carbondata/hive/CarbonObjectInspector.java [hunk 15]
```diff
@@ -60,14 +60,16 @@ public CarbonObjectInspector(final StructTypeInfo rowTypeInfo) {
   }
 
   private ObjectInspector getObjectInspector(final TypeInfo typeInfo) {
-    if (typeInfo.equals(TypeInfoFactory.doubleTypeInfo)) {
+    if (typeInfo.equals(TypeInfoFactory.stringTypeInfo)) {
```
Comment (ajantha-bhat) [inline integration/hive/src/main/java/org/apache/carbondata/hive/CarbonObjectInspector.java]: why this change and why move string up  ? assuming many columns will be string type ?

File: core/src/main/java/org/apache/carbondata/core/cache/dictionary/AbstractColumnDictionaryInfo.java (hunk 10 above)
Comment (jackylk) [reply core/src/main/java/org/apache/carbondata/core/cache/dictionary/AbstractColumnDictionaryInfo.java]: It is used in query flow

File: core/src/main/java/org/apache/carbondata/core/localdictionary/dictionaryholder/DictionaryByteArrayWrapper.java (hunk 11 above)
Comment (jackylk) [reply core/src/main/java/org/apache/carbondata/core/localdictionary/dictionaryholder/DictionaryByteArrayWrapper.java]: removed

File: core/src/main/java/org/apache/carbondata/core/util/CarbonLoadStatisticsImpl.java (hunk 13 above)
Comment (jackylk) [reply core/src/main/java/org/apache/carbondata/core/util/CarbonLoadStatisticsImpl.java]: fixed

File: hadoop/src/main/java/org/apache/carbondata/hadoop/readsupport/impl/DirectDictionaryReadSupport.java (hunk 14 above)
Comment (jackylk) [reply hadoop/src/main/java/org/apache/carbondata/hadoop/readsupport/impl/DirectDictionaryReadSupport.java]: DirectDictionaryReadSupport is removed now

File: integration/hive/src/main/java/org/apache/carbondata/hive/CarbonObjectInspector.java (hunk 15 above)
Comment (jackylk) [reply integration/hive/src/main/java/org/apache/carbondata/hive/CarbonObjectInspector.java]: yes, string should be compared first

File: core/src/main/java/org/apache/carbondata/core/scan/filter/resolver/resolverinfo/visitor/DictionaryColumnVisitor.java (hunk 12 above)
Comment (jackylk) [reply core/src/main/java/org/apache/carbondata/core/scan/filter/resolver/resolverinfo/visitor/DictionaryColumnVisitor.java]: Because I removed `throw QueryExecutionException` in getFilterValues, so it is not needed to catch it

File: core/src/main/java/org/apache/carbondata/core/cache/CacheProvider.java (hunk 9 above)
Comment (jackylk) [reply core/src/main/java/org/apache/carbondata/core/cache/CacheProvider.java]: renamed

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/1333/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/1343/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1356/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/1343/

**Comment by CarbonDataQA1** [comment]
Build Failed with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/1354/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1367/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.1.0, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.1/1351/

**Comment by CarbonDataQA1** [comment]
Build Failed with Spark 2.2.1, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.2/1361/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/1372/

**Comment by QiangCai** [comment]
LGTM
